blogposts/2021/how-we-built-batch-changes.md

```diff
@@ -0,0 +1,197 @@
+that building the product from the ground up, fully understanding and owning the
+code is more important to the long-term success of the project and the team than
+merging a prototype to get something out there as fast as possible. From my
+perspective now this was one of the key decisions that, to this day, helped us
+avoid accumulating large amounts of technical debt, even though we were
+constantly shipping new things.
+
+Only a few months later, though, we were stuck. We realised that what we had
+built and shipped as an alpha to customers just didn't work on a conceptual
+level. In that version, campaigns (now called batch changes) were executed on
+the Sourcegraph instance to produce changes in repositories, which was cool from
+a technical standpoint but it felt clunky, and slow, and adding a new type of
+campaign (to produce a new type of change) required us to add new code. "Meh" is
+a good word to use here.
+
+So at this point we had thrown the prototype away, built the product from
+scratch and had to realise that, well, _this isn't it_. So what did we do?
+
+[We threw half of it away][pr2]. The trigger was this idea: what if we simply
+don't produce changes on the server-side but instead accept _diffs_ through
+the API and, as part of the Sourcegraph CLI, provide a tool to produce those
+diffs wherever you can run the CLI with whatever Docker container you want?
+
+The server-side would then only have concern itself with diffs and how to
+publish them on code hosts as pull and merge requests, how to sync them, how to
+track them, how to rate limit their creation, etc. We wouldn't have to ship a
+new release every time we came up with a new way to produce diffs. And our
+customers that were most interested in making large-scale changes often had
+custom tooling already. With the CLI producing diffs by running any Docker
+container we could give them a place to plug it in — making it large-scale.
+
+This radical change to how the product worked, was, in my opinion, maybe the
+most important one made in the life of Batch Changes. The mental model and the
+abstraction layers it introduced turned out to be incredibly well-fitting. They
+allowed us to go faster and get a _valuable_ tool into the hands of our
+customers even earlier.
+
+So far, so good, right? Yes, except, you know, [naming is
+hard](https://martinfowler.com/bliki/TwoHardThings.html) and while we could've
+made the old name for the feature — _Automation_ — work we decided to rename it.
+Take a look at [PR #3][p3] for the first of many in February 2020 that changed
+"[Aa]utomation" to "[Cc]ampaigns" in our codebase.
+
+What followed were months of fine-tuning, user testing, shipping, tweaking -
+_iterating_. We added more features, fixed a lot of bugs, made things go faster.
+
+But — and by now you should know where this is going — something felt _off_.
+While Campaigns worked and we had customers using it and saying they loved it,
```

Comment:
Just wanted to say that, in this case and the next several that follow, I like that you are using "Campaigns", because you are talking about renaming. So I wouldn't change it in these places. 🙂